### PR TITLE
Cleaned up version of #14

### PR DESCRIPTION
**Pull request by mbechler**
Took the time and cleaned that stuff up quite a bit.

With regards to the issues raised
- JDK Compatibility: Base64 now handled with commons-codec
- Git History: rebased against master, finer-grained and cleaner commits
- License/Copyright: My fault, these were just leftovers from my template. MIT is fine. Do you need me to explicitely state that in the source?
- Tests: Now near complete coverage. Built some infrastucture for that (a bit hackish in some cases).
- Maven profiles: My thinking was that dependencies were getting a bit out of hand, so that saved a few MBs on the final JAR and dependency download when one didn't need some of the more exotic stuff. Also did't see that jenkins repository, thanks. The remaining profiles are build-time switches for bundling different incompatible library versions. These could be removed but without a mechanism for that then one would have to put the alternatives manually on the classpath (testing with the different alternatives would also be harder). 

With regards to #10, maybe an solution could also allow doing that switch at runtime and bundling multiple alternatives. 
I'd probably go with the reflection based approach - in cases where it turns out to be necessary - (like I did in the hibernate payload) as I don't think that build time conflicts will be actually that common and a multi module variant might lead to some duplication.

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/frohoff/ysoserial/pull/20" target="_blank">ReviewNinja

**Comment by frohoff** [comment]
I'll have to take a closer look, but so far this looks great. Thanks a ton for putting in the extra work.

Regarding the licenses: AFAIK it should just inherit the license of the project as a whole as long as there's no other licenses (or "All rights reserved") attached. Make sure your name is somewhere in any code you want to be credited with (gadgets/chains and/or exploits).

Regarding the maven profile stuff: This is probably as good as it will get without a larger overhaul of how this is handled (as per #10), which is at the top of my list to sort out ASAP. Suggestions definitely welcome as usual.

The test stuff you added actually looks great and is very similar to what I had in mind.

I'll probably have a few questions/comments on the commits in the PR, but other than that this looks great. 

Thanks again.

File: src/main/java/ysoserial/payloads/ReleaseableObjectPayload.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+/**
+ * © 2016 AgNO3 Gmbh & Co. KG
+ * All right reserved.
```
Comment (frohoff) [inline src/main/java/ysoserial/payloads/ReleaseableObjectPayload.java]: Looks like one more slipped in ;-)

File: src/main/java/ysoserial/exploit/JBoss.java [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+            }
+        }
+
+        doRun(u, payloadObject, username, password);
+        Utils.releasePayload(args[1], payloadObject);
+    }
+
+
+
+    /**
```
Comment (frohoff) [inline src/main/java/ysoserial/exploit/JBoss.java]: Is it possible to remove all the empty javadoc stuff? Can probably do it post-merge too if it's a pain.

File: src/main/java/ysoserial/exploit/JBoss.java [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+import org.xnio.XnioWorker;
+import org.xnio.ssl.JsseXnioSsl;
+import org.xnio.ssl.XnioSsl;
+
+import ysoserial.payloads.ObjectPayload.Utils;
+
+
+/**
+ * 
+ * An exploitation client for JBoss AS/Wildfly JMX
```
Comment (frohoff) [inline src/main/java/ysoserial/exploit/JBoss.java]: A description of how this works and/or a CVE reference would be ideal here. Presumably this has been reported to JBoss already.

File: src/main/java/ysoserial/exploit/JRMPClient.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import java.net.URLClassLoader;
+import java.net.UnknownHostException;
+
+import javax.net.SocketFactory;
+
+import sun.rmi.transport.TransportConstants;
+import ysoserial.payloads.ObjectPayload.Utils;
+
+
+/**
```
Comment (frohoff) [inline src/main/java/ysoserial/exploit/JRMPClient.java]: A description of how this works would be ideal here, especially if/how it interoperates with other ysoserial components.

File: src/main/java/ysoserial/exploit/JRMPListener.java [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+
+
+/**
+ * @author mbechler
+ *
+ */
+@SuppressWarnings ( {
+    "restriction"
+} )
+public class JRMPListener implements Runnable {
```
Comment (frohoff) [inline src/main/java/ysoserial/exploit/JRMPListener.java]: A description of how this works would be ideal here, especially if/how it interoperates with other ysoserial components.

File: src/main/java/ysoserial/exploit/JSF.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import org.apache.commons.codec.binary.Base64;
+
+import ysoserial.payloads.ObjectPayload.Utils;
+
+
+/**
+ * @author mbechler
+ *
+ */
+public class JSF {
```
Comment (frohoff) [inline src/main/java/ysoserial/exploit/JSF.java]: A description of how this works would be ideal here, especially if/how it interoperates with other ysoserial components.

File: src/main/java/ysoserial/exploit/JenkinsCLI.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+import hudson.remoting.Channel;
+import hudson.remoting.Channel.Mode;
+import hudson.remoting.ChannelBuilder;
+import ysoserial.payloads.ObjectPayload.Utils;
+
+/**
+ * @author mbechler
+ *
+ */
+public class JenkinsCLI {
```
Comment (frohoff) [inline src/main/java/ysoserial/exploit/JenkinsCLI.java]: A description of how this works would be ideal here, especially if/how it interoperates with other ysoserial components. Is this part of CVE-2016-0788 or something separate?

**Comment by frohoff** [comment]
Some short descriptions/notes on how the new `exploit` package and JRMP stuff works would be helpful.

**Comment by mbechler** [comment]
Added some docs for the tooling. That exploit package isn't new, is it? ;) Figured that would be the correct place for tooling delivering playloads.

With regards to JBoss/Wildfly:
Judging from 
http://lists.jboss.org/pipermail/wildfly-dev/2015-November/004581.html
and
https://access.redhat.com/solutions/2045023
they are aware of the problem in general, and as far as my (limited) research goes I haven't found anything that would be directly exploitable in just a base installation (very restrictive classloaders on the exported stuff, also authentiction needed). So this is mainly a tool for people that may want to test their servers.

**Comment by h3xstream** [comment]
/cc @frohoff 
I have a proposition to clean the build while handling the library isolation.

Here is a POC that separate each gadget in modules and would allow simple libraries isolation.
https://github.com/GoSecure/ysoserial/tree/feature/gradle-build

Gradle will bring two main advantages to the projet
 - It will bring minimal configuration for new gadget (compare to XML file required by Maven)
 - Simplify the packaging for library isolation (See [packaging task](https://github.com/GoSecure/ysoserial/blob/feature/gradle-build/ysoserial-cli/build.gradle#L10))

The POC only illustrate the intended structure and packaging. Some code would be needed to dynamically
(See [bootstrap class](https://github.com/GoSecure/ysoserial/blob/feature/gradle-build/ysoserial-cli/src/main/java/ysoserial/BootstrapMain.java#L24)).

Let me know if you have questions.

**Comment by frohoff** [comment]
Thanks again for the contribution and sorry it took so long. Going to try to accept a couple other PRs and do some cleanup before releasing.

**Comment by frohoff** [comment]
Seeing some intermittent flakiness with the PayloadsTest for JRMPClient failing intermittently and was hoping you could take a look if/when you get a chance.

**Comment by mbechler** [comment]
So far haven't been able to reproduce locally. One potential issue I can think of is that, depending on test execution order, the java.rmi.server.useCodebaseOnly property (currently set in Gadgets) is not always set before sun.rmi.server.MarshalInputStream is loaded. Would probably be better to set this directly on the test VM in the pom (also it's not needed otherwise).

**Pull request merged**